backend/retentionpolicy.ts

```diff
@@ -0,0 +1,180 @@
+        await deleteAny(result, deletions[i].schema)
+      }
+      console.log(`Triggered deletion of ${deletions[i].state} ${deletions[i].schema} older than the ${deletions[i].deletionPeriod} days.`)
+    }
+  }
+}
+
+async function deleteAny(reports: Array<ITravel | IExpenseReport | IHealthCareCost>, schema: schemaNames) {
+  let result: any
+  for (let i = 0; i < reports.length; i++) {
+    result = await model(schema).deleteOne({ _id: reports[i]._id })
+    if (result && result.deletedCount == 1) {
+      console.log(`Deleted ${schema} with id ${reports[i]._id} including ${reports[i].history.length} older versions of this ${schema}.`)
+    }
+  }
+}
+
+async function notificationMailForDeletions(
+  daysUntilDeletion: number,
+  deletionPeriodRefunded: number,
+  deletionPeriodApprovedTravel: number,
+  deletionPeriodInWorkReport: number
+) {
+  let notifications: { schema: schemaNames; state: AnyState; deletionPeriod: number }[] = [
+    { schema: 'Travel', state: 'refunded', deletionPeriod: deletionPeriodRefunded },
+    { schema: 'Travel', state: 'approved', deletionPeriod: deletionPeriodApprovedTravel },
+    { schema: 'ExpenseReport', state: 'refunded', deletionPeriod: deletionPeriodRefunded },
+    { schema: 'ExpenseReport', state: 'inWork', deletionPeriod: deletionPeriodInWorkReport },
+    { schema: 'HealthCareCost', state: 'refunded', deletionPeriod: deletionPeriodRefunded },
+    { schema: 'HealthCareCost', state: 'inWork', deletionPeriod: deletionPeriodInWorkReport }
+  ]
+  if (daysUntilDeletion > 0) {
+    try {
+      for (let i = 0; i < notifications.length; i++) {
+        let daysUntilDeletionTemp =
+          daysUntilDeletion < notifications[i].deletionPeriod ? daysUntilDeletion : notifications[i].deletionPeriod
+        console.log(notifications[i].deletionPeriod - daysUntilDeletionTemp)
+        let date = await getDateThreshold(notifications[i].deletionPeriod - daysUntilDeletionTemp)
+        let result = await getForRetentionPolicy(notifications[i].schema, date, notifications[i].state)
+        console.log(
+          `searched for ${notifications[i].schema} with state ${notifications[i].state}  - and deletion in ${daysUntilDeletionTemp} or less days`
+        )
+        if (result.length > 0) {
+          for (let p = 0; p < result.length; p++) {
+            await sendNotificationMails(result[p], notifications[i].deletionPeriod)
+            console.log(`sent notification mail for ${notifications[i].schema} with ID: ${result[p]._id}`)
+          }
+        }
+      }
+    } catch (error) {
+      console.log('Error during sending of notification mails:', error)
+    }
+  } else {
+    console.log('Warning: setting for notification mails before deletion is set to infinite.')
+  }
+}
+
+async function sendNotificationMails(report: ITravel | IExpenseReport | IHealthCareCost, deletionPeriod: number) {
+  if (report) {
+    let owner
+    try {
+      owner = await User.findOne({ _id: report.owner._id })
```

Comment:
Hier kannst du auch .lean() benutzen


```suggestion
      owner = await User.findOne({ _id: report.owner._id }).lean()
```